web/app/components/thread/thread.reducers.js

```diff
@@ -0,0 +1,43 @@
+import { THREAD_SET_COLLAPSE } from './thread.actions';
+import getCollapsedComments from './getCollapsedComments';
+import saveCollapsedComments from './saveCollapsedComments';
+import { siteId, url } from 'common/settings';
+
+const collapsedCommentIds = getCollapsedComments()
+  .map(comment => comment.split('_'))
+  .filter(components => components[0] === siteId && components[1] === url)
+  .map(component => component[2]);
+
+const initialState = collapsedCommentIds.reduce((acc, id) => ({ ...acc, [id]: true }), {});
+
+const persistCollapsedComments = (comment, collapsed) => {
+  const lsCollapsedID = `${siteId}_${url}_${comment.id}`;
+  let collapsedComments = getCollapsedComments();
+
+  if (collapsed) {
+    collapsedComments = [...new Set(collapsedComments.concat(lsCollapsedID))];
+  } else {
+    collapsedComments = collapsedComments.filter(id => id !== lsCollapsedID);
+  }
+
+  saveCollapsedComments(collapsedComments);
+};
+
+export const collapsedThreads = (state = initialState, action) => {
+  switch (action.type) {
+    case THREAD_SET_COLLAPSE: {
+      const currentCollapsed = state[action.comment.id];
+
+      if (action.collapsed !== currentCollapsed) {
+        persistCollapsedComments(action.comment, action.collapsed);
```

Comment:
Reducers must be pure functions. No side effects allowed.

Reply:
We can make custom middleware to persist to localStorage. Using localStorage value for initial value is ok.

Reply:
Agreed, done. If there will be more cases like this - we could try handling all the side effects with sagas